File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/CS/Direct2D/D2DShapes/GeometryShape.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using Microsoft.WindowsAPICodePack.DirectX.Direct2D1;

namespace D2DShapes
{
    internal class GeometryShape : DrawingShape
    {
        private Geometry geometry;
        internal PathGeometry geometryOutlined;
        internal PathGeometry geometrySimplified;
        internal PathGeometry geometryWidened;
        internal Matrix3x2F? worldTransform;

        public GeometryShape(RenderTarget initialRenderTarget, Random random, D2DFactory d2DFactory, D2DBitmap bitmap)
            : base(initialRenderTarget, random, d2DFactory, bitmap)
        {
            coolStrokes = CoinFlip;
            double which = Random.NextDouble();
            if (which < 0.67 || coolStrokes)
                PenBrush = RandomBrush();
            if (!coolStrokes && which > 0.33)
                FillBrush = RandomBrush();
            if (coolStrokes || CoinFlip)
                StrokeStyle = RandomStrokeStyle();
            if (CoinFlip)
                worldTransform = RandomMatrix3x2();
            StrokeWidth = RandomStrokeWidth();
            geometry = RandomGeometry();
            if (coolStrokes || Random.NextDouble() < 0.3)
                ModifyGeometry();
            if (coolStrokes && CoinFlip)
            {
                ModifyGeometry();
            }
        }

        [TypeConverter(typeof(ExpandableObjectConverter))]
        public Geometry Geometry
        {
            get { return geometry; }
            set { geometry = value; }
        }

        private void ModifyGeometry()
        {
            GeometrySink geometrySink;
            double which = Random.NextDouble();
            if (which < 0.33)
            {
                geometryOutlined = d2DFactory.CreatePathGeometry();
                geometrySink = geometryOutlined.Open();
                geometry.Outline(geometrySink, FlatteningTolerance);
                geometrySink.Close();
                geometrySink.Dispose();
                geometry.Dispose();
                geometry = geometryOutlined;
            }
            else if (which < 0.67)
            {
                geometrySimplified = d2DFactory.CreatePathGeometry();
                geometrySink = geometrySimplified.Open();
                geometry.Simplify(
                    CoinFlip
                        ? GeometrySimplificationOption.Lines
                        : GeometrySimplificationOption.CubicsAndLines,
                    geometrySink, 
                    FlatteningTolerance
                    );
                geometrySink.Close();
                geometrySink.Dispose();
                geometry.Dispose();
                geometry = geometrySimplified;
            }
            else
            {
                geometryWidened = d2DFactory.CreatePathGeometry();
                geometrySink = geometryWidened.Open();
                geometry.Widen(
                    RandomStrokeWidth(), //75
                    RandomStrokeStyle(),                    
                    geometrySink,
                    FlatteningTolerance);
                geometrySink.Close();
                geometrySink.Dispose();
                geometry.Dispose();
                geometry = geometryWidened;
            }
        }

        //this could be called recursively
        public Geometry RandomGeometry()
        {
            return RandomGeometry(0);
        }

        public Geometry RandomGeometry(int level)
        {
            Geometry g = null;

            while (g == null)
            {
                double which = Random.NextDouble();

                if (which < 0.20)
                    g = RandomEllipseGeometry();
                else if (which < 0.40)
                    g = RandomRoundRectGeometry();
                else if (which < 0.60)
                    g = RandomRectangleGeometry();
                else if (which < 0.80)
                    g = RandomPathGeometry();
                else if (level < 3)
                    g = RandomGeometryGroup(level + 1);
            }

            if (worldTransform.HasValue)
                g = d2DFactory.CreateTransformedGeometry(g, worldTransform.Value);

            return g;
        }

        private Geometry RandomTransformedGeometry()
        {
            Geometry start, ret;
            start = RandomGeometry();
            ret = d2DFactory.CreateTransformedGeometry(
                start,
                RandomMatrix3x2());
            start.Dispose();
            return ret;
        }
        
        private Geometry RandomGeometryGroup(int level)
        {
            var geometries = new List<Geometry>();
            int count = Random.Next(1, 5);
            for (int i = 0; i < count; i++)
                geometries.Add(RandomGeometry(level));
            GeometryGroup ret = d2DFactory.CreateGeometryGroup(
                Random.NextDouble() < .5 ? FillMode.Winding : FillMode.Alternate,
                geometries);
            foreach (var g in geometries)
            {
                g.Dispose();
            }
            return ret;
        }

        private PathGeometry RandomPathGeometry()
        {
            PathGeometry g = d2DFactory.CreatePathGeometry();
            int totalSegmentCount = 0;
            int figureCount = Random.Next(1, 2);
            using (GeometrySink sink = g.Open())
            {
                for (int f = 0; f < figureCount; f++)
                {
                    int segmentCount = Random.Next(2, 20);
                    AddRandomFigure(sink, segmentCount);
                    totalSegmentCount += segmentCount;
                }
                sink.Close();
            }
            System.Diagnostics.Debug.Assert(g.SegmentCount == totalSegmentCount);
            System.Diagnostics.Debug.Assert(g.FigureCount == figureCount);
            return g;
        }

        private void AddRandomFigure(IGeometrySink sink, int segmentCount)
        {
            previousPoint = null;
            sink.BeginFigure(
                RandomNearPoint(),
                CoinFlip ? FigureBegin.Filled : FigureBegin.Hollow);
            FigureEnd end = CoinFlip ? FigureEnd.Closed : FigureEnd.Closed;
            if (end == FigureEnd.Closed)
                segmentCount--;
            if (CoinFlip)
                for (int i = 0; i < segmentCount; i++)
                    AddRandomSegment(sink);
            else
            {
                double which = Random.NextDouble();
                if (which < 0.33)
                    sink.AddLines(RandomLines(segmentCount));
                else if (which < 0.67)
                    sink.AddQuadraticBeziers(RandomQuadraticBeziers(segmentCount));
                else
                    sink.AddBeziers(RandomBeziers(segmentCount));
            }
            sink.EndFigure(end);
        }

        private IEnumerable<Point2F> RandomLines(int segmentCount)
        {
            var lines = new List<Point2F>();
            for (int i = 0; i < segmentCount; i++)
                lines.Add(RandomNearPoint());
            return lines;
        }

        private IEnumerable<QuadraticBezierSegment> RandomQuadraticBeziers(int segmentCount)
        {
            var beziers = new List<QuadraticBezierSegment>();
            for (int i = 0; i < segmentCount; i++)
                beziers.Add(new QuadraticBezierSegment(
                    RandomNearPoint(),
                    RandomNearPoint()));
            return beziers;
        }

        private IEnumerable<BezierSegment> RandomBeziers(int segmentCount)
        {
            var beziers = new List<BezierSegment>();
            for (int i = 0; i < segmentCount; i++)
                beziers.Add(new BezierSegment(
                    RandomNearPoint(),
                    RandomNearPoint(),
                    RandomNearPoint()));
            return beziers;
        }

        private void AddRandomSegment(IGeometrySink sink)
        {
            double which = Random.NextDouble();
            if (which < 0.25)
                sink.AddLine(RandomNearPoint());
            else if (which < 0.5)
                sink.AddArc(RandomArc());
            else if (which < 0.75)
                sink.AddBezier(RandomBezier());
            else if (which < 1.0)
            sink.AddQuadraticBezier(RandomQuadraticBezier());
        }

        private QuadraticBezierSegment RandomQuadraticBezier()
        {
            return new QuadraticBezierSegment(
                RandomNearPoint(),
                RandomNearPoint());
        }

        private BezierSegment RandomBezier()
        {
            return new BezierSegment(
                RandomNearPoint(),
                RandomNearPoint(),
                RandomNearPoint());
        }

        private ArcSegment RandomArc()
        {
            return new ArcSegment(
                RandomNearPoint(),
                RandomSize(),
                (float)Random.NextDouble() * 360,
                CoinFlip ? SweepDirection.Clockwise : SweepDirection.Counterclockwise,
                CoinFlip ? ArcSize.Large : ArcSize.Small
                );
        }

        private SizeF RandomSize()
        {
            return new SizeF(
                (float)Random.NextDouble() * CanvasWidth,
                (float)Random.NextDouble() * CanvasHeight);
        }

        private RoundedRectangleGeometry RandomRoundRectGeometry()
        {
            return d2DFactory.CreateRoundedRectangleGeometry(
                RandomRoundedRect());
        }

        private RectangleGeometry RandomRectangleGeometry()
        {
            return d2DFactory.CreateRectangleGeometry(
                RandomRect(CanvasWidth, CanvasHeight));
        }

        private EllipseGeometry RandomEllipseGeometry()
        {
            return d2DFactory.CreateEllipseGeometry(
                RandomEllipse());
        }

        protected internal override void ChangeRenderTarget(RenderTarget newRenderTarget)
        {
            PenBrush = CopyBrushToRenderTarget(PenBrush, newRenderTarget);
            FillBrush = CopyBrushToRenderTarget(FillBrush, newRenderTarget);
        }

        protected internal override void Draw(RenderTarget renderTarget)
        {
            Geometry g = geometry;
            if (FillBrush != null)
            {
                renderTarget.FillGeometry(g, FillBrush, null);
            }
            if (PenBrush != null)
            {
                if (StrokeStyle != null)
                {
                    renderTarget.DrawGeometry(g, PenBrush, StrokeWidth, StrokeStyle);
                }
                else
                {
                    renderTarget.DrawGeometry(g, PenBrush, StrokeWidth);
                }
            }
        }

        public override bool HitTest(Point2F point)
        {
            return geometry.FillContainsPoint(point, FlatteningTolerance);
        }

        public override void Dispose()
        {
            if (geometry != null)
                geometry.Dispose();
            geometry = null;
            base.Dispose();
        }
    }
}

File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/VB/Direct2D/D2DShapes/GeometryShape.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.

Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports Microsoft.WindowsAPICodePack.DirectX.Direct2D1

Namespace D2DShapes
	Friend Class GeometryShape
		Inherits DrawingShape
'INSTANT VB NOTE: The variable geometry was renamed since Visual Basic does not allow class members with the same name:
		Private geometry_Renamed As Geometry
		Friend geometryOutlined As PathGeometry
		Friend geometrySimplified As PathGeometry
		Friend geometryWidened As PathGeometry
		Friend worldTransform? As Matrix3x2F

		Public Sub New(ByVal initialRenderTarget As RenderTarget, ByVal random As Random, ByVal d2DFactory As D2DFactory, ByVal bitmap As D2DBitmap)
			MyBase.New(initialRenderTarget, random, d2DFactory, bitmap)
			coolStrokes = CoinFlip
			Dim which As Double = Random.NextDouble()
			If which < 0.67 OrElse coolStrokes Then
				PenBrush = RandomBrush()
			End If
			If (Not coolStrokes) AndAlso which > 0.33 Then
				FillBrush = RandomBrush()
			End If
			If coolStrokes OrElse CoinFlip Then
				StrokeStyle = RandomStrokeStyle()
			End If
			If CoinFlip Then
				worldTransform = RandomMatrix3x2()
			End If
			StrokeWidth = RandomStrokeWidth()
			geometry_Renamed = RandomGeometry()
			If coolStrokes OrElse Random.NextDouble() < 0.3 Then
				ModifyGeometry()
			End If
			If coolStrokes AndAlso CoinFlip Then
				ModifyGeometry()
			End If
		End Sub

		<TypeConverter(GetType(ExpandableObjectConverter))> _
		Public Property Geometry() As Geometry
			Get
				Return geometry_Renamed
			End Get
			Set(ByVal value As Geometry)
				geometry_Renamed = value
			End Set
		End Property

		Private Sub ModifyGeometry()
			Dim geometrySink As GeometrySink
			Dim which As Double = Random.NextDouble()
			If which < 0.33 Then
				geometryOutlined = d2DFactory.CreatePathGeometry()
				geometrySink = geometryOutlined.Open()
                geometry_Renamed.Outline(geometrySink, FlatteningTolerance)
				geometrySink.Close()
				geometrySink.Dispose()
				geometry_Renamed.Dispose()
				geometry_Renamed = geometryOutlined
			ElseIf which < 0.67 Then
				geometrySimplified = d2DFactory.CreatePathGeometry()
				geometrySink = geometrySimplified.Open()
                geometry_Renamed.Simplify(If(CoinFlip, GeometrySimplificationOption.Lines, GeometrySimplificationOption.CubicsAndLines), geometrySink, FlatteningTolerance)
				geometrySink.Close()
				geometrySink.Dispose()
				geometry_Renamed.Dispose()
				geometry_Renamed = geometrySimplified
			Else
				geometryWidened = d2DFactory.CreatePathGeometry()
				geometrySink = geometryWidened.Open()
                geometry_Renamed.Widen(RandomStrokeWidth(), RandomStrokeStyle(), geometrySink, FlatteningTolerance)
				geometrySink.Close()
				geometrySink.Dispose()
				geometry_Renamed.Dispose()
				geometry_Renamed = geometryWidened
			End If
		End Sub

        'this could be called recursively
        Public Function RandomGeometry() As Geometry
            Return RandomGeometry(0)
        End Function

        Public Function RandomGeometry(ByVal level As Integer) As Geometry
            Dim which As Double = Random.NextDouble()
            Dim g As Geometry = Nothing

            While g Is Nothing
                If which < 0.2 Then
                    g = RandomEllipseGeometry()
                ElseIf which < 0.4 Then
                    g = RandomRoundRectGeometry()
                ElseIf which < 0.6 Then
                    g = RandomRectangleGeometry()
                ElseIf which < 0.8 Then
                    g = RandomPathGeometry()
                ElseIf level < 3 Then
                    g = RandomGeometryGroup(level + 1)
                End If
            End While

            If worldTransform.HasValue Then
                g = d2DFactory.CreateTransformedGeometry(g, worldTransform.Value)
            End If
            Return g
        End Function

		Private Function RandomTransformedGeometry() As Geometry
			Dim start, ret As Geometry
			start = RandomGeometry()
			ret = d2DFactory.CreateTransformedGeometry(start, RandomMatrix3x2())
			start.Dispose()
			Return ret
		End Function

        Private Function RandomGeometryGroup(ByVal level As Integer) As Geometry
            Dim geometries = New List(Of Geometry)()
            Dim count As Integer = Random.Next(1, 5)
            For i As Integer = 0 To count - 1
                geometries.Add(RandomGeometry(level))
            Next i
            Dim ret As GeometryGroup = d2DFactory.CreateGeometryGroup(If(Random.NextDouble() < 0.5, FillMode.Winding, FillMode.Alternate), geometries)
            For Each g In geometries
                g.Dispose()
            Next g
            Return ret
        End Function

		Private Function RandomPathGeometry() As PathGeometry
			Dim g As PathGeometry = d2DFactory.CreatePathGeometry()
			Dim totalSegmentCount As Integer = 0
			Dim figureCount As Integer = Random.Next(1, 2)
            Using sink As GeometrySink = g.Open()
                For f As Integer = 0 To figureCount - 1
                    Dim segmentCount As Integer = Random.Next(2, 20)
                    AddRandomFigure(sink, segmentCount)
                    totalSegmentCount += segmentCount
                Next f
                sink.Close()
            End Using
            System.Diagnostics.Debug.Assert(g.SegmentCount = totalSegmentCount)
            System.Diagnostics.Debug.Assert(g.FigureCount = figureCount)
            Return g
        End Function

		Private Sub AddRandomFigure(ByVal sink As IGeometrySink, ByVal segmentCount As Integer)
			previousPoint = Nothing
			sink.BeginFigure(RandomNearPoint(),If(CoinFlip, FigureBegin.Filled, FigureBegin.Hollow))
			Dim [end] As FigureEnd = If(CoinFlip, FigureEnd.Closed, FigureEnd.Closed)
            If [end] = FigureEnd.Closed Then
                segmentCount -= 1
            End If
			If CoinFlip Then
				For i As Integer = 0 To segmentCount - 1
					AddRandomSegment(sink)
				Next i
			Else
				Dim which As Double = Random.NextDouble()
				If which < 0.33 Then
					sink.AddLines(RandomLines(segmentCount))
				ElseIf which < 0.67 Then
					sink.AddQuadraticBeziers(RandomQuadraticBeziers(segmentCount))
				Else
					sink.AddBeziers(RandomBeziers(segmentCount))
				End If
			End If
			sink.EndFigure([end])
		End Sub

		Private Function RandomLines(ByVal segmentCount As Integer) As IEnumerable(Of Point2F)
			Dim lines = New List(Of Point2F)()
			For i As Integer = 0 To segmentCount - 1
				lines.Add(RandomNearPoint())
			Next i
			Return lines
		End Function

		Private Function RandomQuadraticBeziers(ByVal segmentCount As Integer) As IEnumerable(Of QuadraticBezierSegment)
			Dim beziers = New List(Of QuadraticBezierSegment)()
			For i As Integer = 0 To segmentCount - 1
				beziers.Add(New QuadraticBezierSegment(RandomNearPoint(), RandomNearPoint()))
			Next i
			Return beziers
		End Function

		Private Function RandomBeziers(ByVal segmentCount As Integer) As IEnumerable(Of BezierSegment)
			Dim beziers = New List(Of BezierSegment)()
			For i As Integer = 0 To segmentCount - 1
				beziers.Add(New BezierSegment(RandomNearPoint(), RandomNearPoint(), RandomNearPoint()))
			Next i
			Return beziers
		End Function

		Private Sub AddRandomSegment(ByVal sink As IGeometrySink)
			Dim which As Double = Random.NextDouble()
			If which < 0.25 Then
				sink.AddLine(RandomNearPoint())
			ElseIf which < 0.5 Then
				sink.AddArc(RandomArc())
			ElseIf which < 0.75 Then
				sink.AddBezier(RandomBezier())
			ElseIf which < 1.0 Then
			sink.AddQuadraticBezier(RandomQuadraticBezier())
			End If
		End Sub

		Private Function RandomQuadraticBezier() As QuadraticBezierSegment
			Return New QuadraticBezierSegment(RandomNearPoint(), RandomNearPoint())
		End Function

		Private Function RandomBezier() As BezierSegment
			Return New BezierSegment(RandomNearPoint(), RandomNearPoint(), RandomNearPoint())
		End Function

		Private Function RandomArc() As ArcSegment
			Return New ArcSegment(RandomNearPoint(), RandomSize(), CSng(Random.NextDouble()) * 360,If(CoinFlip, SweepDirection.Clockwise, SweepDirection.CounterClockwise),If(CoinFlip, ArcSize.Large, ArcSize.Small))
		End Function

		Private Function RandomSize() As SizeF
			Return New SizeF(CSng(Random.NextDouble()) * CanvasWidth, CSng(Random.NextDouble()) * CanvasHeight)
		End Function

		Private Function RandomRoundRectGeometry() As RoundedRectangleGeometry
			Return d2DFactory.CreateRoundedRectangleGeometry(RandomRoundedRect())
		End Function

		Private Function RandomRectangleGeometry() As RectangleGeometry
			Return d2DFactory.CreateRectangleGeometry(RandomRect(CanvasWidth, CanvasHeight))
		End Function

		Private Function RandomEllipseGeometry() As EllipseGeometry
			Return d2DFactory.CreateEllipseGeometry(RandomEllipse())
		End Function

		Protected Friend Overrides Sub ChangeRenderTarget(ByVal newRenderTarget As RenderTarget)
			PenBrush = CopyBrushToRenderTarget(PenBrush, newRenderTarget)
			FillBrush = CopyBrushToRenderTarget(FillBrush, newRenderTarget)
		End Sub

		Protected Friend Overrides Sub Draw(ByVal renderTarget As RenderTarget)
			Dim g As Geometry = geometry_Renamed
			If FillBrush IsNot Nothing Then
				renderTarget.FillGeometry(g, FillBrush, Nothing)
			End If
			If PenBrush IsNot Nothing Then
				If StrokeStyle IsNot Nothing Then
					renderTarget.DrawGeometry(g, PenBrush, StrokeWidth, StrokeStyle)
				Else
					renderTarget.DrawGeometry(g, PenBrush, StrokeWidth)
				End If
			End If
		End Sub

		Public Overrides Function HitTest(ByVal point As Point2F) As Boolean
			Return geometry_Renamed.FillContainsPoint(point, FlatteningTolerance)
		End Function

		Public Overrides Sub Dispose()
			If geometry_Renamed IsNot Nothing Then
				geometry_Renamed.Dispose()
			End If
			geometry_Renamed = Nothing
			MyBase.Dispose()
		End Sub
	End Class
End Namespace
